experiment/simulation/js/gate.js

```diff
@@ -0,0 +1,317 @@
+import { registerGate, jsPlumbInstance } from "./main.js";
+import { setPosition } from "./layout.js";
+import { decoderTest, displayTest, computeAnd, computeOr, computeXor, computeXnor, computeNand, computeNor } from "./validator.js";
+
+'use strict';
+
+export let gates = {}; // Array of gates
+window.numComponents = 0;
+export function clearGates() {
+
+    for (let gateId in gates) {
+        delete gates[gateId];
+    }
+
+    gates = {};
+}
+
+export class Gate {
+    constructor(type) {
+        this.type = type;
+        this.id = type + "-" + window.numComponents++;
+        this.positionX = 0;
+        this.positionY = 0;
+        this.isConnected = false;
+        this.inputPoints = [];
+        this.outputPoints = [];
+        this.inputs = []; // List of input gates
+        this.output = null; // Output value
+        this.isInput = false;
+        this.isOutput = false;
+        this.name = null;
+    }
+    setId(id) {
+        this.id = id;
+    }
+    addInput(gate) {
+        this.inputs.push(gate);
+    }
+    removeInput(gate) {
+        let index = this.inputs.indexOf(gate);
+        if (index > -1) {
+            this.inputs.splice(index, 1);
+        }
+    }
+    updatePosition(id) {
+        this.positionY = window.scrollY + document.getElementById(id).getBoundingClientRect().top // Y
+
+        this.positionX = window.scrollX + document.getElementById(id).getBoundingClientRect().left // X
+    }
+    setName(name) {
+        this.name = name;
+    }
+
+    generateComponent() {
+        let component = "";
+
+        switch (this.type) {
+            case "Input":
+                component = `<div class="high" id= ${this.id} ><a ondblclick="setInput(event)">1</a><p> ${this.name}  </p></div>`;
+                this.output = true;
+                this.isInput = true;
+                break;
+            case "Output":
+                component = `<div class="output" id= ${this.id}><a></a><p>  ${this.name}  </p></div>`;
+                this.isOutput = true;
+                break;
+            default:
+                component = `<div class="drag-drop logic-gate ${this.type.toLowerCase()}" id= ${this.id
+                    }></div>`;
+        }
+        return component;
+    }
+
+    registerComponent(workingArea, x = 0, y = 0) {
+        // get width of working area
+        const width = document.getElementById(workingArea).offsetWidth;
+        const height = document.getElementById(workingArea).offsetHeight;
+        let scale = 900;
+        let yScale = 800;
+        x = (x / scale) * width;
+        y = (y / yScale) * height;
+
+        const el = document.getElementById(this.id);
+        el.style.left = x + "px";
+        el.style.top = y + "px";
+
+        if (this.type != "Input" && this.type != "Output") {

```

Comment:
Is this part needed?

Reply:
yes, explained in previous experiments review